src/App.php

```diff
@@ -1063,25 +1072,50 @@ function () {
      */
     protected function outputResponseUnsafe(string $data, array $headersNew): void
     {
+        $isSse = $this->response->getHeaderLine('Content-Type') === 'text/event-stream';
+
+        if (headers_sent() && $isSse) {
+            echo $data;
```

Comment:
there is no streaming in PSR-7 response?

otherwise we should build it/write to it as for normal response

also, what is the point of building PSR-7 response here when we echo it a few lines later?

Reply:
There is many discussion on this topic :
https://github.com/slimphp/Slim/issues/2481
https://github.com/laminas/laminas-diactoros/issues/16

I see many suggests to use callback stream but in this case we have to make changes at JsSse level, Slim have a different approach which uses this : https://github.com/slimphp/Slim-Psr7/blob/master/src/NonBufferedBody.php

In Atk JsSse is blocking and we cannot return the response, while JsSse is in callback execution, so we can take this KISS approach and output following request to output data or we can use something like this : https://github.com/phly/psr7examples/blob/master/src/CallbackStream.php and execute the Sse in returned stream, which is IMHO better but we have to set special response directly in JsSse, and callback firing will be done when response will be emitted.

I like the second one, but i try to make as less changes as possible, to avoid conflict with other open PR.

Reply:
the nonBufferedBody seems not usefull at all

> CallbackStream

seems all to address read-once concept, which is not what we actually need

> KISS approach and output following request to output data

with custom Response class (that we can detect using instanceof) this seems the best to me and individual SSE responses there

I think it is ok for now if we keep the SSE impl. minimal like now and implement it only when there is a valid usecase and we actually know what the usecase needs.

One major question still: if target use still need to extent App (to override `emitResponse`, othewise App will echo), why is PSR-7 helpfull to have in App if the target use can override `outputResponseUnsafe` in a simillar fashion?

If I do not missed something, having `response` in an app is not usefull at all (as it is built once per request anyway, we do not need it as a place to add a data to) :)

Reply:
We should integrate https://github.com/abbadon1334/atk4-psr7/tree/master/src into this repo. The `PSR7App` should be transformed to a trait with `outputResponseUnsafe` method and `protected getDefaultsForPsr7Response` method to get the needed default properties which should be merged with the defaults in an user App class constructor.

Reply:
agree, but now if we remove the response and remove the last added method `emitResponse`, we can output directly the response in `outputResponseUnsafe` as we do before.
At this point if someone needs PSR7 implementation, we can avoid copy/paste the whole method `outputResponse` just to change 2 lines, now is all much clear with `outputResponseUnsafe`.

About useful or not useful, Psr7 if implemented correctly will allow integration of Atk4, in every framework nearly without touching it, but practically, the emitter of response is not needed, Psr7 is for Request/Response standardization to allow implementation of routing + middleware.

About integration, I think I need to find a correct way to implement JsSse after that we speak about integration, at now except for usage of NonBufferedBody, which equally breaks the output stream, this PR and atk4-psr7 are at the same point.

Reply:
> If I do not missed something, having response in an app is not usefull at all (as it is built once per request anyway, we do > not need it as a place to add a data to) :)

Rethinking about your affirmation, i think with Response we have another way to simplify some code in App.
if we remove normalization and other things we do to set/check headers in App response and redirect all directly to Response to build it? At the end we have only to emit the response in place of recover status + headers and build and emit

Reply:
Well, it is all quite handly and well written there... The normalization and diffing there is better than what it can be then. So I prefer to keep it, maybe one day once all other ugly things are fixed :)

Reply:
Totally agree, but FYI in Nyholm Response when you call withHeader internally normalize the header, clearly doing what we do in App https://github.com/Nyholm/psr7/blob/fafc2faf8d49b979f922fe416dc8805462dc1fe8/src/MessageTrait.php#L59-L89

Reply:
> internally normalize the header, clearly doing what we do in App

then go ahead with it, no need to have the code twice!

we however need header unset for https://github.com/atk4/ui/blob/develop/src/App.php#L126, but that is solveable if we never allow DI for the response class (and I belive we do not have to or I did not get the usecase)